src/responder/multiResponder.ts

```diff
@@ -0,0 +1,237 @@
+     *   This replacement rate is set by the nodes.
+     *   Parity: 12.5%: https://github.com/paritytech/parity-ethereum/blob/master/miner/src/pool/scoring.rs#L38
+     *   Geth: 10% default : https://github.com/ethereum/go-ethereum/wiki/Command-Line-Options --txpool.pricebump
+     */
+    constructor(
+        signer: ethers.Signer,
+        private readonly gasEstimator: GasPriceEstimator,
+        private readonly transactionTracker: TransactionTracker,
+        public readonly maxConcurrentResponses: number = 12,
+        public readonly replacementRate: number = 1.13
+    ) {
+        super(signer);
+        this.txMined = this.txMined.bind(this);
+        this.broadcast = this.broadcast.bind(this);
+    }
+
+    // we do some async setup
+    private async setup() {
+        if (!this.queue) {
+            const address = await this.signer.getAddress();
+            const nonce = await this.provider.getTransactionCount(address);
+            this.queue = new GasQueue([], nonce, this.replacementRate, this.maxConcurrentResponses);
+            this.chainId = (await this.provider.getNetwork()).chainId;
+        }
+    }
+
+    public async startResponse(appointmentId: string, responseData: IEthereumResponseData) {
+        try {
+            await this.setup();
+            if (this.queue.depthReached()) {
+                throw new Error(`Cannot add to queue. Max queue depth ${this.queue.maxQueueDepth} reached.`);
+            }
+
+            // form a queue item request
+            const abiInterface = new ethers.utils.Interface(responseData.contractAbi);
+            const data = abiInterface.functions[responseData.functionName].encode(responseData.functionArgs);
+            const txIdentifier = new PisaTransactionIdentifier(
+                this.chainId,
+                data,
+                responseData.contractAddress,
+                new BigNumber(0),
+                new BigNumber(EthereumResponder.GAS_LIMIT)
+            );
+            const idealGas = await this.gasEstimator.estimate(responseData);
+            const request = new GasQueueItemRequest(txIdentifier, idealGas, responseData);
+
+            // add the queue item to the queue, since the queue is ordered this may mean
+            // that we need to replace some transactions on the network. Find those and
+            // broadcast them
+            const replacedQueue = this.queue.add(request);
+            const replacedTransactions = replacedQueue.difference(this.queue);
+            this.queue = replacedQueue;
+
+            await Promise.all(replacedTransactions.map(this.broadcast));
+        } catch (doh) {
+            if (doh instanceof QueueConsistencyError) logger.error(doh.stack!);
+            else {
+                logger.error(`Unexpected error trying to respond for: ${appointmentId}.`);
+                if (doh.stack) logger.error(doh.stack);
+                else logger.error(doh);
+            }
+        }
+    }
```

Comment:
:+1: This is a nice idea - one way to implement this would be for queue items to know whether they have been issued to the network and so could be replaced for free. I'm not a big fan of this as it means recording more information in the queue - and another function to finalise all the requests. Another way would be to store up requests, then try to decide the optimum order in which to add them to the queue to minimise replacements. I like this even less as it means inferring behaviour of the queue from outside the queue.